### PR TITLE
Refactor: Causal Mask Update and Prepare for Generate

### PR DESCRIPTION
**Pull request by vasqu**
Refactor for the functions in the title. Also a small fix on half inference with cache.

### CODE REVIEW

File: src/transformers/models/mamba2/__init__.py [hunk 1]
```diff
@@ -55,4 +55,4 @@
 else:
     import sys
 
-    sys.modules[__name__] = _LazyModule(__name__, globals()["__file__"], _import_structure, module_spec=__spec__)
\ No newline at end of file
+    sys.modules[__name__] = _LazyModule(__name__, globals()["__file__"], _import_structure, module_spec=__spec__)
```
Comment (vasqu) [inline src/transformers/models/mamba2/__init__.py:58]: Make style added this.

File: src/transformers/models/mamba2/configuration_mamba2.py [hunk 2]
```diff
@@ -295,4 +295,4 @@ def _attention_layers_idx_validation(self):
             raise ValueError(
                 "`attention_layers_idx` has out-of-range indices, "
                 f"got {self.attention_layers_idx}, but expected indices in {list(range(self.num_hidden_layers))}"
-            )
\ No newline at end of file
+            )
```
Comment (vasqu) [inline src/transformers/models/mamba2/configuration_mamba2.py:298]: Make style added this.

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 3]
```diff
@@ -1373,7 +1373,7 @@ def slow_forward(self, hidden_states, cache):
             )
 
             # Subsequent output
-            y = torch.einsum("bhpn,bn->bhp", cache.ssm_states[self.layer_idx], C)
+            y = torch.einsum("bhpn,bn->bhp", cache.ssm_states[self.layer_idx].to(C.dtype), C)
```
Comment (vasqu) [inline src/transformers/models/mamba2/modeling_mamba2.py:1376]: We save the cache in fp32 but we compute in the current type, meaning we need to reconvert on computation :D

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 4]
```diff
@@ -1724,10 +1724,8 @@ def forward(
             inputs_embeds = self.embeddings(input_ids)
         hidden_states = inputs_embeds
 
-        # Force cache to be our custom hybrid one, something in the generation module incorrectly overwrites it...
-        if (past_key_values is None and use_cache) or not isinstance(
-            past_key_values, HybridMamba2AttentionDynamicCache
-        ):
+        # We allow empty caches on initial forward
+        if past_key_values is None and use_cache:
```
Comment (vasqu) [inline src/transformers/models/mamba2/modeling_mamba2.py:1728]: Moved to prepare for generation to reconvert to our format

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 5]
```diff
@@ -1801,34 +1801,80 @@ def forward(
...
+
+        # TODO: check if this is compatible with this custom cache format
+        if self._attn_implementation == "sdpa" and not output_attentions:
+            if AttentionMaskConverter._ignore_causal_mask_sdpa(
+                attention_mask,
+                inputs_embeds=input_tensor,
+                past_key_values_length=past_seen_tokens,
+                is_training=self.training,
+            ):
+                return None
```
Comment (vasqu) [inline src/transformers/models/mamba2/modeling_mamba2.py:1834]: Not sure if this is correct since llama does not do it for a static cache. No idea why tho...

File: src/transformers/models/mamba2/modeling_mamba2.py [hunk 6]
```diff
@@ -1868,65 +1914,56 @@ def set_input_embeddings(self, new_embeddings):
...
-                input_ids = input_ids[:, past_length:]
-
-            # TODO: skipping the usual 3. as we do not use a sliding window or use a define a max_cache_len
-        else:
+            if inputs_embeds is not None:  # Exception 1
+                input_ids = input_ids[:, -cache_position.shape[0] :]
+            elif input_ids.shape[1] != cache_position.shape[0]:  # Default case (the "else", a no op, is Exception 2)
+                input_ids = input_ids[:, cache_position]
+
+        # Force cache to be our custom hybrid one, something in the generation module incorrectly overwrites it...
```
Comment (vasqu) [inline src/transformers/models/mamba2/modeling_mamba2.py:1940]: Some bug imo in transformers main; hard to find where it happens and it should happen from their side to fix it.

**Comment by vasqu** [comment]
Relevant PR in main https://github.com/huggingface/transformers/pull/31898
